backends/concrete-cpu/implementation/src/implementation/fast_keyswitch.rs

```diff
@@ -0,0 +1,272 @@
+
+    let mut output_glwe = GlweCiphertext::new(output_glwe_c.as_mut_slice(), ggsw.out_glwe_params());
+
+    trivial_packing(input_glwe.as_mut_view(), input);
+
+    external_product(
+        output_glwe.as_mut_view(),
+        ggsw,
+        input_glwe.as_view(),
+        fft,
+        stack,
+    );
+
+    output_glwe.as_view().sample_extract(out, 0);
+
+    (input_glwe_c, output_glwe_c)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::c_api::types::tests::to_generic;
+    use crate::implementation::encrypt::encrypt_constant_ggsw;
+    use crate::implementation::fast_keyswitch::fast_keyswitch;
+    use crate::implementation::fft::Fft;
+    use crate::implementation::types::{
+        CsprngMut, DecompParams, GgswCiphertext, GlweCiphertext, GlweParams, GlweSecretKey,
+        LweCiphertext, LweSecretKey,
+    };
+    use concrete_csprng::generators::{RandomGenerator, SoftwareRandomGenerator};
+    use concrete_csprng::seeders::Seed;
+    use dyn_stack::DynStack;
+    use std::mem::MaybeUninit;
+
+    use super::trivial_packing;
+
+    fn init_sk(sk: &mut [u64], csprng: CsprngMut) {
+        let len = sk.len();
+        LweSecretKey::new(sk, len).fill_with_new_key(csprng)
+    }
+
+    #[test]
+    fn test_trivial_packing() {
+        let in_lwe_dim = 100;
+
+        let out_glwe_params = GlweParams {
+            dimension: 2,
+            polynomial_size: 512,
+        };
+
+        let polynomial_size = out_glwe_params.polynomial_size;
+
+        let mut csprng = SoftwareRandomGenerator::new(Seed(0));
+        let mut csprng = to_generic(&mut csprng);
+
+        let mut sk = vec![0; in_lwe_dim];
+
+        init_sk(&mut sk, csprng.as_mut());
+
+        let sk_lwe = LweSecretKey::new(sk.as_slice(), in_lwe_dim);
+
+        let mut sk_glwe = vec![0; out_glwe_params.lwe_dimension()];
+
+        sk_glwe[0..in_lwe_dim].copy_from_slice(sk_lwe.as_view().data);
+
+        let sk_glwe = GlweSecretKey::new(sk_glwe.as_slice(), out_glwe_params);
+
+        let in_ct = 1_u64 << 60;
+
+        let mut in_lwe_ct = vec![0; in_lwe_dim + 1];
+
+        let mut in_lwe_ct = LweCiphertext::new(in_lwe_ct.as_mut_slice(), in_lwe_dim);
+
+        sk_lwe.encrypt_lwe(in_lwe_ct.as_mut_view(), in_ct, 0.0, csprng.as_mut());
+
+        let mut out_glwe_ct = vec![0; (out_glwe_params.dimension + 1) * polynomial_size];
+
+        let mut out_glwe_ct = GlweCiphertext::new(out_glwe_ct.as_mut_slice(), out_glwe_params);
+
+        trivial_packing(out_glwe_ct.as_mut_view(), in_lwe_ct.as_view());
+
+        let decrypted_lwe = sk_lwe.decrypt_lwe(in_lwe_ct.as_view());
+
+        assert!(decrypted_lwe.abs_diff(in_ct) < 1_u64 << 30);
+
+        let decrypted_glwe = sk_glwe.decrypt_glwe(out_glwe_ct.as_view());
+
+        assert_eq!(decrypted_lwe, decrypted_glwe[0]);
+    }
+
+    #[test]
+    fn fast_ks() {
+        let decomp_params = DecompParams {
+            level: 3,
+            base_log: 10,
+        };
+
+        let in_lwe_dim = 300;
```

Comment:
Same as above, better to have something realistic like 650

Reply:
Done